File: Python/Array/122. Best Time to Buy and Sell Stock II.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        cur_cost = float('-inf')
        profit = 0
        prev_profit = 0
        for price in prices:
            if prev_profit > price + cur_cost:
                profit += prev_profit
                prev_profit = 0
                cur_cost = -price
                continue
            cur_cost = max(-price, cur_cost)
            prev_profit = max(prev_profit, price + cur_cost)
            # elif cur_cost < -price:
            #     cur_cost = -price
            # elif prev_profit < price + cur_cost:
            #     prev_profit = price + cur_cost
        profit += prev_profit
        return profit

File: CPP/array/122. Best Time to Buy and Sell Stock II.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int cur_cost = -INT_MAX;
        int prev_profit = 0;
        for (int price: prices){
            if (price + cur_cost < prev_profit){
                profit += prev_profit;
                prev_profit = 0;
                cur_cost = -price;
            }
            else if (-price > cur_cost){
                cur_cost = -price;
            }
            else if (price + cur_cost > prev_profit){
                prev_profit = price + cur_cost;
            }
            // cout << "cur price:" << price << ", profit:" << profit << ", cur_cost:" << cur_cost << ", prev_profit:" << prev_profit << endl;
        }
        profit += prev_profit;
        return profit;
    }
};